コメントを付けられるようにする

gaki273 opened the pull request.

itagakisan committed:
`reports`テーブルをデータベースに追加 Model追加

itagakisan committed:
コントローラファイルを追加 Controller追加

[目的]
- 日報機能に関するWebリクエストを処理するため

[変更内容]
- 各コントローラファイルを追加

itagakisan committed:
日報の一覧を表示する機能を追加 機能追加

[目的]
- 日報の一覧を表示

[変更内容]
- ルートファイルにて、日報機能のルーティングを定義
- `ReportsController`に`index`アクションを追加し、全ての日報を取得してビューに渡す処理を実装
- 取得した日報はID順にソートし、ページネーションを適用

itagakisan committed:
日報の一覧画面を作成 View追加

[目的]
- 作成された日報の一覧を表示

[変更内容]
- `reports/index_html.erb`を追加し、日報の一覧を表示するページを作成
- `reports/_report.html.erb`パーシャルを追加し、個別のレポート情報を表示するレイアウトを定義
- 一覧画面にページネーションを導入
- 新規日報作成ページへのリンクと、各日報の詳細ページへのリンクを追加

itagakisan committed:
`new`アクションを追加 機能追加

[目的]
- 新規日報作成ページを表示する

[変更内容]
- `ReportsController`に`new`アクションを追加
- `Report.new`で、ビューで利用する空の日報オブジェクトを生成

itagakisan committed:
新規日報作成画面を追加 View追加

[目的]
- 新規日報作成画面の表示

[変更内容]
- 新規日報作成画面を実装
- タイトルと本文の入力フォーム、およびエラーメッセージ表示機能を実装
- 日報一覧へ戻るリンクを追加

itagakisan committed:
新規日報作成機能の実装 機能追加

[目的]
- フォームから送信された新しい日報を作成し、データベースに保存するロジックを実装。

[変更内容]
- `ReportsController`に`create`アクションを追加
- フォームから送信されたパラメータを使用して日報を新規作成し、データベースに保存する処理を実装
- 保存成功時には詳細ページへリダイレクトし、失敗時には新規作成フォームを再表示するよう実装
- Strong Parametersを使用して、安全にパラメータを受け取るための`report_params`メソッドを定義

itagakisan committed:
日報の詳細画面を表示する機能の追加 機能追加

[目的]
- 特定の日報の詳細ページを表示する機能を追加

[変更内容]
- `ReportsController`に`show`アクションを追加
- `before_action`で`set_book`メソッドを定義し、`show`アクションの前に該当の日報を取得する共通処理を実装

itagakisan committed:
日報の詳細画面を作成 View追加

[目的]
- 特定の日報の詳細ページを表示する

[変更内容]
- `/reports/show.html.erb`を追加
- `_report`パーシャルで日報の詳細を表示
- 編集ページへのリンク、一覧ページへ戻るリンク、および削除ボタンを設置

itagakisan committed:
日報の編集機能を追加 機能追加

[目的]
- 特定の日報を編集できるようにする

[変更内容]
- `ReportsController`に`edit`と`update`アクションを追加
- `edit`アクションで編集フォームを表示するよう実装
- `update`アクションでフォームから送信されたデータで日報を更新する処理を実装
- 更新成功時は詳細ページへリダイレクトし、失敗時は編集ページを再表示
- `before_action`に`:edit`, `:update`を追加し、対象の日報を事前に取得するようにした

itagakisan committed:
日報の編集画面を追加 View追加

[目的]
- 特定の日報を編集する画面を表示

[変更内容]
- `/reports/edit.html.erb`を追加
- `_form`パーシャルで、日報の編集フォームを表示
- 詳細ページへのリンクと、一覧ページへ戻るリンクを設置

itagakisan committed:
日報の削除機能の実装 機能追加

[目的]
- 特定の日報を削除する

[変更内容]
- `ReportsController`に`destroy`アクションを追加
- `destroy`アクションで日報を削除し、日報一覧ページにリダイレクトする処理を実装
- `before_action`に`:destroy`を追加し、削除対象の日報を事前に取得するようにした

itagakisan committed:
ナビゲーションバーに日報一覧ページへのリンクを追加 View修正

[目的]
- 全てのページから日報一覧へアクセスできるようにする。

[変更内容]
- `application.html.erb`のナビゲーションバーに、日報一覧ページへのリンクを追加

itagakisan committed:
翻訳ファイルに日報の日本語を追加 修正

[目的]
- 日報モデルとその属性に日本語の翻訳を適用する

[変更内容]
- `Report`モデル名を「日報」として定義
- `Report`モデルの`title`属性を「タイトル」、`content`属性を「内容」として定義

itagakisan committed:
メソッド名を修正 修正

itagakisan committed:
日報の作成者表示のためにマイグレーションファイルを追加 機能追加

[目的]
- 日報とその作成者を紐づける

[変更内容]
-  `reports`テーブルに`user_id`カラム（null: false）を追加
- `user_id`カラムにインデックスを追加
- `reports`テーブルと`users`テーブル間に外部キー制約を追加

itagakisan committed:
UserとReportの関連付け 機能追加

[目的]
- 作成者をわかるようにする

[変更内容]
- UserモデルとReportモデルにアソシエーションを定義
- reports_controllerのcreateアクションを修正

itagakisan committed:
日報に投稿者名を表示 機能追加

[目的]
- 日報の作成者を表示する

[変更内容]
- 日報の部分テンプレートを修正し、投稿者名を表示するようにした

itagakisan committed:
日報の編集・削除機能を投稿者のみに制限 機能追加

[目的]
- 他者が編集・削除をできないようにする

[変更内容]
- `reports/show.html.erb`を修正、`if`文にて日報の投稿者本人にのみ編集・削除ボタンが表示されるようにした

itagakisan committed:
`comments'テーブルをデータベースに追加 Model追加 Model追加

itagakisan committed:
コメント機能の関連付け 機能追加

[目的]
- `Book`と`Report`にコメントを投稿できるようにする

[変更内容]
- `Book`モデルに、`Comment`モデルとのポリモーフィック関連を定義
- `Report`モデルに、`Comment`モデルとのポリモーフィック関連を定義
- `User`モデルに、`Comment`モデルとの関連を定義

itagakisan committed:
コメント機能のルーティングを定義 機能追加

[目的]
- 本と日報にコメントを投稿・削除できるようにするため

[変更内容]
- `books`リソースにネストして`comments`のルーティングを追加
- `reports`リソースにネストして`comments`のルーティングを追加

itagakisan committed:
コントローラファイルを追加 Controller追加

[目的]
- コメント機能に関するWebリクエストを処理するため

[変更内容]
- `Comments`の各コントローラファイルを追加

itagakisan committed:
コメントを投稿できるように`create`アクションを追加 機能追加

[目的]
- 本や日報にコメントを投稿できるようにする

[変更内容]
- `create`アクションを定義
- before_actionでコメント対象のモデル（Book or Report）を判定

itagakisan committed:
日報の`show`アクションに処理を追加 機能追加

[目的]
- 日報詳細ページでコメントを投稿・表示できるようにするため

[変更内容]
- `reports_controller`の`show`アクションに、コメント対象のインスタンス変数を追加

itagakisan committed:
コメント投稿フォームと一覧表示のパーシャルを作成 View追加

[目的]
- 日報詳細ページでコメントの投稿と閲覧ができるようにする

[変更内容]
- コメント投稿フォームのパーシャルを作成
- コメント一覧表示のパーシャルを作成
- 日報詳細ページにコメントフォームと一覧表示のパーシャルを設置
- タイムゾーンを東京に設定

itagakisan committed:
コメントのスタイル調整

itagakisan committed:
コメントの削除機能の追加 機能追加

[目的]
- 投稿したコメントを削除する。

[変更内容]
- `CommentsController`に`destroy`アクションを追加
- コメントを削除できるように処理を追加

itagakisan committed:
コメントの削除ボタンの追加 機能追加

[目的]
- 投稿したコメントを削除をする

[変更内容]
- コメントのパーシャルに削除ボタンを追加

itagakisan committed:
翻訳ファイルにコメントの日本語を追加 修正

[目的]
- コメント機能の日本語対応

[変更内容]
- `ja.yml` に `post` と `comment` のキーを追加
- `translation_ja.yml` に `comment` モデルとその属性の翻訳を追加

itagakisan committed:
コメント機能の表示の修正 修正

[目的]
- コメントの件数を表示
- 投稿ボタンをわかりやすくする

[変更内容]
- コメント一覧のヘッダーにコメント数を表示するように修正

itagakisan committed:
コメントのフォームのスタイルを調整 修正

[目的]
- コメント投稿フォームのレイアウトを整える

[変更内容]
- コメントフォームのパーシャルを`div`で囲み、スタイル用のクラスを付与
- `scaffolds.css`にコメントフォーム用のスタイルを追加

itagakisan committed:
本へのコメント機能を追加 機能追加

[目的]
- 本の詳細ページでコメントを投稿・閲覧できるようにする

[変更内容]
- `books_controller`の`show`アクションに、コメント機能で利用するインスタンス変数`@commentable`を追加

itagakisan committed:
本へのコメントの各パーシャルを追 View追加

[目的]
- 本の詳細ページでコメントの投稿と閲覧ができるようにする

[変更内容]
- 書籍詳細ページにコメントフォームと一覧表示のパーシャルを設置

itagakisan committed:
本へのコメントの各パーシャルを追 View追加

[目的]
- 本の詳細ページでコメントの投稿と閲覧ができるようにする

[変更内容]
- コメント一覧のパーシャルで、コメント数を`@commentable`変数で取得するように修正

itagakisan committed:
自分の投稿したコメントのみを削除できるように修正 修正

[目的]
- 自分の投稿したコメントのみを削除できるようにする

[変更内容]
- ログインユーザーが投稿者の場合のみ削除ボタンを表示するように修正

itagakisan committed:
入力必須化 修正

[目的]
- 日報やコメントのから投稿を防ぐ

[変更内容]
- コメント投稿フォームの`content`に `required: true` を追加
- 日報投稿フォームの`title`と`content`に `required: true` を追加

itagakisan committed:
翻訳ファイルに投稿者の日本語を追加 修正

[目的]
- 日報の投稿者ラベルを日本語で表示する

[変更内容]
- `translation_ja.yml` に日報の `poster` の翻訳を追加

itagakisan committed:
`Report`モデルにバリデーションを追加 機能追加

[目的]
- 日報のタイトルと内容が空の状態で保存されるのを防ぐ

[変更内容]
- `Report`モデルに`title`と`content`の`presence: true`バリデーションを追加した

itagakisan committed:
日報の表示（一覧・詳細）の修正 修正

[目的]
- 日報の一覧表示と詳細表示をわかりやすくする

[変更内容]
- 日報のパーシャルに投稿者と投稿日時を表示するヘッダーを追加
- 詳細ページでのみヘッダーが表示されるように修正
- ヘッダー用のスタイルを追加

itagakisan committed:
コメント一覧のスタイルを調整 修正

[目的]
- 日報の一覧表示をわかりやすくする

[変更内容]
- コメントアイテムのpaddingを調整し、内側の余白を広げた

yuuu commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,45 @@
+    respond_to do |format|
+      if @comment.save
+        format.html { redirect_to @commentable, notice: t('controllers.common.notice_create', name: Comment.model_name.human) }
+        format.json { render :show, status: :created, location: @comment }
```
今回の仕様であれば `JSON` をレスポンスする必要はないので `respond_to` は使わない実装としましょう 👍

gaki273 replied:
かしこまりました！
`respond_to`を使わない実装になるよう修正いたします！

yuuu commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,45 @@
+  end
+
+  def destroy
+    @comment = Comment.find(params[:id])
```
この取得方法だと他人のコメントを削除できてしまうおそれがあります

gaki273 replied:
かしこまりました！
現在、Webセキュリティのプラクティスを進めており、このあたりの実装は良くないなと感じておりました。
コメントを削除できないように修正いたします。

yuuu commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,45 @@
+        format.html { redirect_to @commentable, notice: t('controllers.common.notice_create', name: Comment.model_name.human) }
+        format.json { render :show, status: :created, location: @comment }
+      else
+        format.html { render :new, status: :unprocessable_entity }
```
バリデーションエラーが発生したとき、この記載方法だと画面が表示されないと思われます。
実際に試してみて、表示されない場合は見直しをお願いします 🙏

gaki273 replied:
かしこまりました！
HTML5の必須チェックが効かなかった場合、空のコメントが投稿され、エラーが表示されないことを確認いたしました！
正しくバリデーションエラーが動作するように修正いたします。

yuuu commented on app/controllers/reports_controller.rb:
```diff
@@ -0,0 +1,64 @@
+    respond_to do |format|
+      if @report.save
+        format.html { redirect_to report_url(@report), notice: t('controllers.common.notice_create', name: Report.model_name.human) }
+        format.json { render :show, status: :created, location: @report }
```
日報についてもJSONを応答する必要はないので `respond_to` は使わない方向で修正をお願いします 🙏

gaki273 replied:
かしこまりました！
コメント同様、`respond_to`を使わないよう修正いたします。

yuuu commented on app/controllers/reports_controller.rb:
```diff
@@ -0,0 +1,64 @@
+  end
+
+  def destroy
+    @report.destroy
```
日報についても他人の日報を編集・削除できてしまうように見えます。
View側でガードをかけていても、curl等で直接リクエストされるとすり抜けてしまいます。

gaki273 replied:
ご指摘の通りです。
他人の日報を編集・削除出来ないように修正いたします。

yuuu commented on app/helpers/comments_helper.rb:
```diff
@@ -0,0 +1,4 @@
+# frozen_string_literal: true
+
+module CommentsHelper
```
使う予定のないヘルパーは削除しておきましょう 🙏

gaki273 replied:
かしこまりました！

yuuu commented on app/views/comments/_comment.html.erb:
```diff
@@ -0,0 +1,25 @@
+    <div id="<%= dom_id comment %>">
+      <div class="comment-header">
+        <p>
+          <%= comment.user&.name.present? ?comment.user.name : comment.user&.email %>
```
ここはいずれも `user` に紐づく処理なので、 `User` モデルにメソッドを追加してそこで判別するようにしましょう

gaki273 replied:
かしこまりました！
`User`モデルにメソッドを追加し、判別するような実装に修正いたします。

yuuu commented on test/models/comment_test.rb:
```diff
@@ -0,0 +1,9 @@
+
+require 'test_helper'
+
+class CommentTest < ActiveSupport::TestCase
```
このあたりもテストとして機能しないファイルという認識ですので削除しておきましょう

gaki273 replied:
かしこまりました！

itagakisan committed:
不要なファイルを削除 削除

itagakisan committed:
`respond_to`の削除、`notice`メッセージの修正 リファクタリング

[目的]
- 不必要な`JSON`形式のレスポンスを削除する。
- 正しい翻訳が出力されるようにする。

[変更内容]
- `CommentsController`の`create`、`destroy`アクションから`respond_to`を削除
- `ReportsController`の`create`、`update`、`destroy`アクションから`respond_to`を削除
- `CommentsController`の`destroy`アクションの`notice`メッセージを修正

itagakisan committed:
コメント・レポートの削除機能を修正 修正

[目的]
- 他人のコメント・レポートを削除出来てしまう問題を修正

[変更内容]
- `CommentsController`の`destroy`アクションで自分のコメントのみ削除可能に変更
- `ReportsController`の`destroy`アクションで自分のレポートのみ削除可能に変更

itagakisan committed:
コメント機能の修正 修正

[目的]
- コメントの内容が空で投稿する際に、バリデーションエラーが発生せず、エラーメッセージも表示されない状況の改善

[変更内容]
- `Comment`モデルに`content`の必須バリデーションを追加
- コメント作成エラー時の処理を修正

itagakisan committed:
コメント・レポート機能の修正 修正

[目的]
- ユーザー名表示の重複ロジックを共通化する

[変更内容]
- `User`モデルに`display_name`メソッドを追加
- コメント・レポート表示でユーザー名表示ロジックを`display_name`メソッドに統一

itagakisan committed:
レポート編集画面からレポート一覧画面へのパスを正しく設定 修正

yuuu commented on app/controllers/comments_controller.rb:
```diff
@@ -7,26 +7,20 @@ def create
+      redirect_to @commentable, notice: t('controllers.common.notice_create', name: Comment.model_name.human)
+    else
+      flash[:alert] = @comment.errors.full_messages.join
+      redirect_to @commentable
```
ここは `redirect_to` してしまうと、ユーザーがフォームに入力していた情報が全て消えてしまうという副作用があります。今回の場合、バリデーションエラーが起こるとしたらコメントが未入力の場合のみなので顕在化はしないかもしれませんが、今後他の入力項目が増えた際等に使いづらさが生じます。

Railsのお作法通り `render` できないか、再度検討をお願いします 🙏

gaki273 replied:
かしこまりました。
`render`での実装を考えていましたが、エラーが多発し断念しておりました。
修正いたします！

yuuu commented on app/controllers/reports_controller.rb:
```diff
@@ -0,0 +1,53 @@
+  end
+
+  def update
+    if @report.update(report_params)
```
他人の日報を削除することはできなくなりましたが、更新はできてしまうように見えます 👀

gaki273 replied:
`update`アクションへの対応を忘れておりました！
修正いたします！

itagakisan committed:
レポート更新処理の権限制御 修正

[目的]
- 他のユーザーのレポートを不正に更新されることを防ぐ

[変更内容]
- `update`メソッドで`current_user.reports.find`を使用して所有者チェックを追加

itagakisan committed:
レポート作成エラー処理 修正

[目的]
- エラー時のフォームデータを保持するため

[変更内容]
- エラー時に`redirect_to`ではなく`render`を使用してページ遷移を防止
- case文でコメント対象の型に応じた処理を実装
- コメント対象に応じて適切なインスタンス変数(`@book`, `@report`)を設定
- `422 Unprocessable Entity`ステータスコードを返却

yuuu commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,45 @@
+# frozen_string_literal: true
+
+class CommentsController < ApplicationController
```
このコントローラは複数箇所で「本と日報の分岐」が存在しており、もし今後コメントができる対象が増えた場合その分岐は更に大きくなっていく懸念があります。

対策として、本へのコメントと日報へのコメントでコントローラを分けられないか検討をお願いします 🙏

gaki273 replied:
かしこまりました。
各コメントでコントローラを分け、共通の処理に関してはモジュール化することで修正いたしました！

itagakisan committed:
`Commentable concern`追加 機能追加

[目的]
- コメント機能の共通処理をモジュール化する

[変更内容]
- `Commentable concern`モジュールを追加

itagakisan committed:
`Commetable concern`モジュールに、CRUD機能を追加 機能追加

[目的]
- 本へのコメントと日報のコメントのコントローラを分けた時の共通の処理をモジュール化する

[変更内容]
- `create`アクションを追加
- `destroy`アクションを追加
- `comment_params`メソッドを追加

itagakisan committed:
`Commets`コントローラを削除、`Book`と`Report`のコメント機能に分岐 機能追加

[目的]
- 今後のコメント機能の拡張を考慮し、各コメントごとにコントローラを作成する

[変更内容]
- `BookCommentsController`に`Commentable concern`を追加
- `ReportCommentsController`に`Commentable concern`を追加
- 各コントローラーに`set_commentable`メソッドを実装

itagakisan committed:
ルーティングをコントローラに合わせて修正 修正

[目的]
- 専用のコメントコントローラーにルーティングするため

[変更内容]
- `books`の`comments`を`book_comments`コントローラーにルーティング
- `reports`の`comments`を`report_comments`コントローラーにルーティング

yuuu commented on app/controllers/report_comments_controller.rb:
```diff
@@ -0,0 +1,13 @@
+# frozen_string_literal: true
+
+class ReportCommentsController < ApplicationController
```
コントローラですが、 `Reports::CommentsController` としてディレクトリ構成もそれに合わせると綺麗ですね。
今のコントローラ名だと `ReportComment` というモデルが存在するかのように見えてしまいます。

gaki273 replied:
かしこまりました！
コントローラ名とディレクトリ構成を修正いたします！

itagakisan committed:
コメントコントローラーの名前空間化 リファクタリング

[目的]
- コントローラーを名前空間で整理する

[変更内容]
- `BookCommentsController`と`ReportCommentsController`を削除
- `app/controller`に`books`と`reports`ディレクトリを追加
- 各コメントコメントコントローラファイルを追加
- ルーティングをbooks/commentsとreports/commentsに変更

yuuu commented on app/controllers/concerns/commentable.rb:
```diff
@@ -0,0 +1,31 @@
+      redirect_to @commentable, notice: t('controllers.common.notice_create', name: Comment.model_name.human)
+    else
+      flash[:alert] = @comment.errors.full_messages.join
+      render "#{@commentable.model_name.route_key}/show", status: :unprocessable_entity
```
ReportやBookに依存する処理をそれぞれのCommentsControllerに寄せたので、ここの `@commentable.model_name.route_key` についても同様に対処しましょう 🙏

gaki273 replied:
かしこまりました。
おっしゃる通りですね。気づいていませんでした💦
修正いたします！

gaki273 replied:
各`CommentsController`で、コメント作成時のエラーの際に呼び出す`on_save_error`メソッドを作成しました。このメソッドで、ビュー用の変数の設定と対応する`show`テンプレートをレンダリングするよう修正いたしました。
このメソッドを追加したことで、エラー時のレンダリング方法が変わったため、暗黙的なpartialレンダリングでテンプレートミスエラーが発生するようになりました。
そのため、`render partial:`を使って明示的に`partial`ファイルとローカル変数を指定する必要があり、そのような変更となっております。

itagakisan committed:
コメント作成時のエラーハンドリングをリファクタリング リファクタリング

[目的]
- エラーハンドリングを各コントローラーで制御できるようにする

[変更内容]
- `Books::CommentsController`に`on_save_error`メソッドを追加
- `Reports::CommentsController`に`on_save_error`メソッドを追加
- `on_save_error`メソッドでは、各変数の設定とテンプレート表示を行う
- `Commentable concern`で`on_save_error`を呼び出すように変更

itagakisan committed:
`partial`レンダリングを明示的に変更 修正

[目的]
- `partial`レンダリングを明示的にする

[変更内容]
- `books/show`で`partial`指定に変更
- reports/showで`partial`指定に変更
